### PR TITLE
Added additional OpenId Introspection/Revocation endpoints and PKCE to the OpenID Settings and Application UI pages

### PR DESCRIPTION
**Pull request by mcalasa**
Addresses these issues: 
Closes https://github.com/OrchardCMS/OrchardCore/issues/10803
Closes https://github.com/OrchardCMS/OrchardCore/issues/7138#issuecomment-723453045

### CODE REVIEW

File: src/OrchardCore.Cms.Web/.config/dotnet-tools.json [hunk 1]
```diff
@@ -0,0 +1,12 @@
+{
```
Comment (kevinchalet) [inline src/OrchardCore.Cms.Web/.config/dotnet-tools.json:1]: @sebastienros or @jtkech should likely take a look to ensure it's something that should be included in the repo 😃

File: src/OrchardCore.Modules/OrchardCore.OpenId/Configuration/OpenIdServerConfiguration.cs [hunk 2]
```diff
@@ -90,6 +90,17 @@ public void Configure(OpenIddictServerOptions options)
             if (settings.UserinfoEndpointPath.HasValue)
             {
                 options.UserinfoEndpointUris.Add(new Uri(settings.UserinfoEndpointPath.Value, UriKind.Relative));
+
```
Comment (kevinchalet) [inline src/OrchardCore.Modules/OrchardCore.OpenId/Configuration/OpenIdServerConfiguration.cs:93]: Please remove the extra lines.

File: src/OrchardCore.Modules/OrchardCore.OpenId/Configuration/OpenIdServerConfiguration.cs [hunk 3]
```diff
@@ -146,6 +157,15 @@ public void Configure(OpenIddictServerOptions options)
...
             }
 
+            if (settings.RequireProofKeyForCodeExchange)
+            {
+                options.RequireProofKeyForCodeExchange = true;
+            }
+            else
+            {
+                options.RequireProofKeyForCodeExchange= false;
+            }
```
Comment (kevinchalet) [inline src/OrchardCore.Modules/OrchardCore.OpenId/Configuration/OpenIdServerConfiguration.cs:167]:
```suggestion
            options.RequireProofKeyForCodeExchange = settings.RequireProofKeyForCodeExchange;
```

File: src/OrchardCore.Modules/OrchardCore.OpenId/Controllers/ApplicationController.cs [hunk 4]
```diff
@@ -174,7 +176,8 @@ public async Task<IActionResult> Create(CreateOpenIdApplicationViewModel model,
                 RedirectUris = model.RedirectUris,
                 Roles = model.RoleEntries.Where(x => x.Selected).Select(x => x.Name).ToArray(),
                 Scopes = model.ScopeEntries.Where(x => x.Selected).Select(x => x.Name).ToArray(),
-                Type = model.Type
+                Type = model.Type,
+                RequirePkce = model.RequirePkce
```
Comment (kevinchalet) [inline src/OrchardCore.Modules/OrchardCore.OpenId/Controllers/ApplicationController.cs:180]: For consistency, we should probably name these properties `RequireProofKeyForCodeExchange`.

File: src/OrchardCore.Modules/OrchardCore.OpenId/Views/Application/Create.cshtml [hunk 5]
```diff
@@ -165,114 +176,114 @@
 </form>
 <script at="Foot">
     //<![CDATA[
-    window.onload = function () {
+    window.onload = function() {
```
Comment (kevinchalet) [inline src/OrchardCore.Modules/OrchardCore.OpenId/Views/Application/Create.cshtml:179]: Please undo these formatting changes.

**Review by kevinchalet — COMMENTED** [review]
Thanks for your PR! ❤️

File: src/OrchardCore.Modules/OrchardCore.OpenId/Configuration/OpenIdServerConfiguration.cs (hunk 2 above)
Comment (mcalasa) [reply src/OrchardCore.Modules/OrchardCore.OpenId/Configuration/OpenIdServerConfiguration.cs]: Removed the extra lines per your request

File: src/OrchardCore.Modules/OrchardCore.OpenId/Controllers/ApplicationController.cs (hunk 4 above)
Comment (mcalasa) [reply src/OrchardCore.Modules/OrchardCore.OpenId/Controllers/ApplicationController.cs]: Yup, make sense.  This has been done.

File: src/OrchardCore.Modules/OrchardCore.OpenId/Views/Application/Create.cshtml (hunk 5 above)
Comment (mcalasa) [reply src/OrchardCore.Modules/OrchardCore.OpenId/Views/Application/Create.cshtml]: My bad, last time I'm using the 'Format Document' in VS.  This has been completed.

File: src/OrchardCore.Cms.Web/.config/dotnet-tools.json [hunk 6]
```diff
@@ -0,0 +1,12 @@
...
+  "version": 1,
+  "isRoot": true,
+  "tools": {
+    "dotnet-ef": {
+      "version": "6.0.5",
+      "commands": [
```
Comment (Skrypt) [inline src/OrchardCore.Cms.Web/.config/dotnet-tools.json:7]: Remove this file.

File: src/OrchardCore.Cms.Web/Properties/launchSettings.json [hunk 7]
```diff
@@ -3,7 +3,7 @@
     "windowsAuthentication": false,
     "anonymousAuthentication": true,
     "iisExpress": {
-      "applicationUrl": "http://localhost:2918",
+      "applicationUrl": "http://localhost:2919",
```
Comment (Skrypt) [inline src/OrchardCore.Cms.Web/Properties/launchSettings.json:6]: Revert changes in this file.

File: src/OrchardCore.Modules/OrchardCore.OpenId/Configuration/OpenIdServerConfiguration.cs [hunk 8]
```diff
@@ -89,7 +89,15 @@ public void Configure(OpenIddictServerOptions options)
             }
             if (settings.UserinfoEndpointPath.HasValue)
             {
-                options.UserinfoEndpointUris.Add(new Uri(settings.UserinfoEndpointPath.Value, UriKind.Relative));
+                options.UserinfoEndpointUris.Add(new Uri(settings.UserinfoEndpointPath.Value, UriKind.Relative));                
+            }
+            if (settings.IntrospectionEndpointPath.HasValue)
```
Comment (hishamco) [inline src/OrchardCore.Modules/OrchardCore.OpenId/Configuration/OpenIdServerConfiguration.cs:94]: Add new line after each `if` block

File: src/OrchardCore.Modules/OrchardCore.OpenId/OpenIdApplicationExtensions.cs [hunk 9]
```diff
@@ -179,6 +182,31 @@ public static async Task UpdateDescriptorFromSettings(this IOpenIdApplicationMan
                 descriptor.Permissions.Remove(OpenIddictConstants.Permissions.ResponseTypes.CodeIdTokenToken);
                 descriptor.Permissions.Remove(OpenIddictConstants.Permissions.ResponseTypes.CodeToken);
             }
+            if (model.AllowIntrospectionEndpoint)
```
Comment (hishamco) [inline src/OrchardCore.Modules/OrchardCore.OpenId/OpenIdApplicationExtensions.cs:185]: Same here

**Review by hishamco — CHANGES_REQUESTED** [review]
Please fix formatting and remove the achnages in the OC.Cms.Web as @Skrypt mentioned

File: src/OrchardCore.Modules/OrchardCore.OpenId/Views/Application/Create.cshtml [hunk 10]
```diff
@@ -115,6 +115,17 @@
...
             </div>
+        </div>
+         <div class="mb-3" >
+             <div class="mb-3">
+                <div class="form-check">
+                    <input type="checkbox" class="form-check-input" asp-for="RequireProofKeyForCodeExchange" checked="@Model.RequireProofKeyForCodeExchange">
+                    <label class="form-check-label" asp-for="RequireProofKeyForCodeExchange">@T["Require Proof Key for Code Exchange"]</label>
+                </div>
+                 <div class="hint">
+                    @T["Ensure that the client/client library supports PKCE before enabling this option."]
```
[the remark this answers is not in the document]
Comment (kevinchalet) [reply src/OrchardCore.Modules/OrchardCore.OpenId/Views/Application/Create.cshtml:126]: For things to work correctly, the user must check that the **client application** *and* the **client library** it uses under the hood must support PKCE. Maybe it should be reworded to clarify this?

[the remark this answers is not in the document]
Comment (mcalasa) [reply src/OrchardCore.Modules/OrchardCore.OpenId/Views/Application/Create.cshtml:126]: Ah got it.  So I'll change it to read **"Ensure that the client application and OAuth or OIDC library being used supports PKCE before enabling this option."**
This more clear?

[the remark this answers is not in the document]
Comment (kevinchalet) [reply src/OrchardCore.Modules/OrchardCore.OpenId/Views/Application/Create.cshtml:126]: Sounds great 😃

[the remark this answers is not in the document]
Comment (mcalasa) [reply src/OrchardCore.Modules/OrchardCore.OpenId/Views/Application/Create.cshtml]: Done, thanks for the input.

**Review by hishamco — APPROVED** [review]
From other than OpenId, everything seems good to me, @kevinchalet your approval is very important for this one

File: src/OrchardCore.Modules/OrchardCore.OpenId/Configuration/OpenIdServerConfiguration.cs [hunk 11]
```diff
@@ -79,17 +79,30 @@ public void Configure(OpenIddictServerOptions options)
...
+
             if (settings.TokenEndpointPath.HasValue)
             {
                 options.TokenEndpointUris.Add(new Uri(settings.TokenEndpointPath.Value, UriKind.Relative));
             }
+
             if (settings.UserinfoEndpointPath.HasValue)
             {
-                options.UserinfoEndpointUris.Add(new Uri(settings.UserinfoEndpointPath.Value, UriKind.Relative));
+                options.UserinfoEndpointUris.Add(new Uri(settings.UserinfoEndpointPath.Value, UriKind.Relative));
```
Comment (kevinchalet) [inline src/OrchardCore.Modules/OrchardCore.OpenId/Configuration/OpenIdServerConfiguration.cs:95]: Looks like some unwanted whitespaces were added at the end of this line?

File: src/OrchardCore.Modules/OrchardCore.OpenId/Services/OpenIdServerService.cs [hunk 12]
```diff
@@ -78,7 +78,9 @@ private OpenIdServerSettings GetSettingsFromContainer(ISite container)
                 AuthorizationEndpointPath = "/connect/authorize",
                 LogoutEndpointPath = "/connect/logout",
                 TokenEndpointPath = "/connect/token",
-                UserinfoEndpointPath = "/connect/userinfo"
+                UserinfoEndpointPath = "/connect/userinfo",
+                IntrospectionEndpointPath = "/connect/introspect",
+                RevocationEndpointPath = "/connect/revoke",
```
Comment (kevinchalet) [inline src/OrchardCore.Modules/OrchardCore.OpenId/Services/OpenIdServerService.cs:83]: Please remove the unnecessary `,` and update the code comment to indicate the introspection and revocation endpoints are enabled by default too.

**Review by kevinchalet — APPROVED** [review]
A few minor changes should be done prior to merging it, but it's almost ready 😃 

Thanks again for your contribution!

**Comment by mcalasa** [comment]
@kevinchalet No problem, glad I can help.

**Pull request merged**